packages/upset/src/components/Columns/BookmarkStar.tsx

```diff
@@ -1,28 +1,134 @@
+import {
+  FC,
+  MouseEvent,
+  useContext,
+  useMemo,
+  useState,
+} from 'react';
+import { useRecoilValue } from 'recoil';
+import { ProvenanceContext } from '../Root';
 import { dimensionsSelector } from '../../atoms/dimensionsAtom';
 import translate from '../../utils/transform';
-import { bookmarkedColorPalette } from '../../atoms/config/currentIntersectionAtom';
+import {
+  bookmarkedColorPalette,
+  bookmarkSelector,
+  nextColorSelector,
+} from '../../atoms/config/currentIntersectionAtom';
 
 type Props = {
-    row: Row;
-}
+  row: Row;
+};
 
+const BASE_OPACITY = 0.0;
+const HOVERED_OPACITY = 0.5;
+const BOOKMARKED_OPACITY = 1.0;
+
+/**
+ * BookmarkStar component renders a star icon that can be bookmarked.
+ * It changes its color and opacity based on the bookmark and hover states.
+ *
+ * @component
+ * @param {Props} props - The properties object.
+ * @param {Object} props.row - The row data for the current item.
+ * @returns {JSX.Element} The rendered BookmarkStar component.
+ *
+ * @example
+ * <BookmarkStar row={row} />
+ */
 export const BookmarkStar: FC<Props> = ({ row }) => {
   const dimensions = useRecoilValue(dimensionsSelector);
   const colorPallete = useRecoilValue(bookmarkedColorPalette);
+  const nextColor = useRecoilValue(nextColorSelector);
+  const bookmarks = useRecoilValue(bookmarkSelector);
+  const { actions } = useContext(ProvenanceContext);
+
+  const [hovered, setHovered] = useState(false);
+  const bookmarked = useMemo(() => bookmarks.find((b) => b.id === row.id), [
+    bookmarks,
+    row.id,
+  ]);
+
+  const color = useMemo(() => (bookmarked ? colorPallete[row.id] : nextColor), [
+    colorPallete,
+    row.id,
+    bookmarked,
+    nextColor,
+  ]);
+
+  /**
+   * Calculates the opacity value based on the bookmark and hover states.
+   *
+   * @returns {number} The opacity value which can be one of the following:
+   * - `BOOKMARKED_OPACITY` if the item is bookmarked.
+   * - `HOVERED_OPACITY` if the item is hovered.
+   * - `BASE_OPACITY` if neither condition is met.
+   *
+   * @param {boolean} bookmarked - Indicates if the item is bookmarked.
+   * @param {boolean} hovered - Indicates if the item is hovered.
+   */
+  const opacity = useMemo(() => {
+    if (bookmarked) {
+      return BOOKMARKED_OPACITY;
+    }
+
+    if (hovered) {
+      return HOVERED_OPACITY;
+    }
+
+    return BASE_OPACITY;
+  }, [bookmarked, hovered]);
+
+  const handleMouseEnter = () => {
+    setHovered(true);
+  };
+
+  const handleMouseLeave = () => {
+    setHovered(false);
+  };
+
+  /**
+   * Handles the click event on the bookmark star icon.
+   * Stops the event from propagating and adds a bookmark with the specified details.
+   *
+   * @param event - The mouse event triggered by clicking the bookmark star icon.
+   */
+  const handleClick = (event: MouseEvent<SVGGElement, MouseEvent>) => {
+    event.stopPropagation();
+    if (bookmarked) {
+      actions.removeBookmark(bookmarked);
+      return;
```

Comment:
I think better style here is `if...else`; and you won't need brackets since both cases are one-liners.

Reply:
Resolved in https://github.com/visdesignlab/upset2/pull/416/commits/5d0cb81f3996cd6522b148fb81a1b64887e7efee